crypto/bn/bn_gcd.c

```diff
@@ -642,9 +642,9 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
 
     for (i = 0; i < m; i++) {
         /* conditionally flip signs if delta is positive and g is odd */
-        cond = (-delta >> (8 * sizeof(delta) - 1)) & g->d[0] & 1
+        cond = ((unsigned int)-delta >> (8 * sizeof(delta) - 1)) & g->d[0] & 1
             /* make sure g->top > 0 (i.e. if top == 0 then g == 0 always) */
-            & (~((g->top - 1) >> (sizeof(g->top) * 8 - 1)));
+            & (~((unsigned int)(g->top - 1) >> (sizeof(g->top) * 8 - 1)));
```

Comment:
Just a thought: would `sizeof(g->top)` be better written as `sizeof(unsigned int)` now given that we have an explicit cast? `g->top` is an `int` at the moment. We are implicitly assuming that sizeof(int) == sizeof(unsigned int), which is probably fairly safe on all sane platforms. If however we change the type of `g->top` to something else with a different size this could introduce a subtle bug. Same for the instance below.

Reply:
Hm.... interesting. It depends on what we would change the top type to. Let's assume it would be something bigger than unsigned int. In that case this should produce a compilation warning because the shift would be bigger than the number of bits in unsigned int. So I do not think the problem would be unnoticed. However if we do sizeof(unsigned int) there instead - this will NOT! be a correct code either if g->top is bigger because we are cutting off the top bits from top and they might be significant. And this will be unnoticed because no warnings will be produced and it will be also unlikely to produce test failures unless we will have really big bignums.

Now let's examine the case where we change the type of top to something smaller than unsigned int (unlikely but anyway). Although the shift will be smaller than the size of unsigned int, it will be still correct because the lowest bit after the shift will still indicate whether `g->top - 1` was negative or not.

So I prefer keeping it as is.

Reply:
ok

Reply:
The C standard mandates that sizeof(int) == sizeof(unsigned int) so there is no problem here.